ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/debug/BreakpointStorageImpl.java

```diff
@@ -0,0 +1,260 @@
+            LOG.warning("Local storage is not supported. Breakpoints won't be preserved.");
+            this.readAllBreakpointMarker = promiseProvider.resolve(null);
+        } else {
+            this.readAllBreakpointMarker = getReadAllBreakpointMarker();
+            preserve();
+            clearOutdatedRecords();
+        }
+    }
+
+    @Override
+    public void addAll(final List<Breakpoint> breakpoints) {
+        readAllBreakpointMarker.then(onFulfilled -> {
+            BreakpointStorageImpl.this.breakpoints.addAll(breakpoints);
+
+            if (storage != null) {
+                preserve();
+            }
+        });
+    }
+
+    @Override
+    public void add(final Breakpoint breakpoint) {
+        readAllBreakpointMarker.then(onFulfilled -> {
+            BreakpointStorageImpl.this.breakpoints.add(breakpoint);
+
+            if (storage != null) {
+                preserve();
+            }
+        });
+    }
+
+    @Override
+    public void delete(final Breakpoint breakpoint) {
+        readAllBreakpointMarker.then(onFulfilled -> {
+            BreakpointStorageImpl.this.breakpoints.removeIf(
+                    b -> b.getLineNumber() == breakpoint.getLineNumber() && b.getPath().equals(breakpoint.getPath())
+            );
+
+            if (storage != null) {
+                preserve();
+            }
+        });
+    }
+
+    @Override
+    public void deleteAll(final List<Breakpoint> breakpoints) {
+        readAllBreakpointMarker.then(onFulfilled -> {
+            for (Breakpoint breakpoint : breakpoints) {
+                BreakpointStorageImpl.this.breakpoints.removeIf(
+                        b -> b.getLineNumber() == breakpoint.getLineNumber() && b.getPath().equals(breakpoint.getPath())
+                );
+            }
+
+            if (storage != null) {
+                preserve();
+            }
+        });
+    }
+
+    @Override
+    public void clear() {
+        readAllBreakpointMarker.then(onFulfilled -> {
+            BreakpointStorageImpl.this.breakpoints.clear();
+
+            if (storage != null) {
+                preserve();
+            }
+        });
+    }
+
+    @Override
+    public Promise<List<Breakpoint>> readAll() {
+        return readAllBreakpointMarker.then(onFulfilled -> breakpoints, onRejected -> Collections.emptyList());
+    }
+
+    private void preserve() {
+        List<BreakpointDto> breakpoints2save = new LinkedList<>();
+
+        for (Breakpoint breakpoint : breakpoints) {
+            breakpoints2save.add(dtoFactory.createDto(BreakpointDto.class)
+                                           .withLocation(dtoFactory.createDto(LocationDto.class)
+                                                                   .withTarget(breakpoint.getPath())
+                                                                   .withLineNumber(breakpoint.getLineNumber())));
+        }
+
+        String storageKey = LOCAL_STORAGE_BREAKPOINTS_KEY + appContext.getWorkspaceId();
```

Comment:
You can define a class field called STORAGE_KEY with `LOCAL_STORAGE_BREAKPOINTS_KEY + appContext.getWorkspaceId()`, cause you have to construct it twice in this class.